src/Filecoin/Address.cpp

```diff
@@ -16,165 +15,283 @@ namespace TW::Filecoin {
+        return std::nullopt;
+    }
+    if (decoded.size() < checksumSize) {
+        return std::nullopt;
+    }
+
+    Data payload(decoded.begin(), decoded.end() - checksumSize);
+    if (!Address::isValidPayloadSize(type, payload.size())) {
+        return std::nullopt;
+    }
+
+    Data actualChecksum(decoded.end() - checksumSize, decoded.end());
+    Data expectedChecksum = calculateChecksum(type, actorID, payload);
+    if (actualChecksum != expectedChecksum) {
+       return std::nullopt;
+    }
+
+    return Address(type, actorID, payload);
+}
+
+bool Address::isValid(const std::string& string) {
+    return Address::fromString(string).has_value();
+}
+
+bool Address::isValid(const Data& encoded) {
+    return Address::fromBytes(encoded).has_value();
+}
+
+Address Address::secp256k1Address(const PublicKey& publicKey) {
+    Data payload = Hash::blake2b(publicKey.bytes, 20);
+    return Address(Type::SECP256K1, 0, payload);
+}
+
+Address Address::delegatedAddress(const PublicKey& publicKey) {
+    if (publicKey.type != TWPublicKeyTypeSECP256k1Extended) {
+        throw std::invalid_argument("Ethereum::Address needs an extended SECP256k1 public key.");
+    }
+
+    const auto data = publicKey.hash({}, Hash::HasherKeccak256, true);
+    Data payload(data.end() - 20, data.end());
+
+    return Address(Type::DELEGATED, ETHEREUM_ADDRESS_MANAGER_ACTOR_ID, payload);
+}
+
+Address Address::delegatedAddress(uint64_t actorID, const Data& payload) {
+    return Address(Type::DELEGATED, actorID, payload);
 }
 
 Address::Address(const std::string& string) {
-    if (!isValid(string))
+    auto addr = Address::fromString(string);
+    if (!addr) {
         throw std::invalid_argument("Invalid address data");
-
-    Type type = parseType(string[1]);
-    // First byte is type
-    bytes.push_back(static_cast<uint8_t>(type));
-    if (type == Type::ID) {
-        uint64_t id = std::stoull(string.substr(2));
-        while (id >= 0x80) {
-            bytes.push_back(((uint8_t)id) | 0x80);
-            id >>= 7;
-        }
-        bytes.push_back((uint8_t)id);
-        return;
     }
 
-    Data decoded;
-    if (!Base32::decode(string.substr(2), decoded, BASE32_ALPHABET_FILECOIN))
+    type = addr->type;
+    actorID = addr->actorID;
+    payload = std::move(addr->payload);
+}
+
+Address::Address(const Data& encoded) {
+    auto addr = Address::fromBytes(encoded);
+    if (!addr) {
         throw std::invalid_argument("Invalid address data");
-    uint8_t payloadSize = Address::payloadSize(type);
+    }
 
-    bytes.insert(bytes.end(), decoded.data(), decoded.data() + payloadSize);
+    type = addr->type;
+    actorID = addr->actorID;
+    payload = std::move(addr->payload);
 }
 
-Address::Address(const Data& data) {
-    if (!isValid(data)) {
+Address::Address(Type type_, uint64_t actorID_, const Data& payload_):
+    type(type_),
+    actorID(actorID_),
+    payload(payload_) {
+    if (!isValidPayloadSize(type_, payload_.size())) {
```

Comment:
let's not use `_` as a suffix, it's reserved by the standard library - use `this->` to make a clear separation, or rename member field as `m_fieldName;`